modules/ngx_http_upstream_vnswrr_module/ngx_http_upstream_vnswrr_module.c

```diff
@@ -0,0 +1,576 @@
+            rrp->tried[i] = 0;
+        }
+
+        ngx_http_upstream_rr_peers_unlock(peers);
+
+        rc = ngx_http_upstream_get_vnswrr_peer(pc, vnsp);
+
+        if (rc != NGX_BUSY) {
+            return rc;
+        }
+
+        ngx_http_upstream_rr_peers_wlock(peers);
+    }
+
+    ngx_http_upstream_rr_peers_unlock(peers);
+
+    pc->name = peers->name;
+
+    return NGX_BUSY;
+}
+
+
+static ngx_int_t
+ngx_http_upstream_get_rr_peer(ngx_http_upstream_rr_peers_t *peers,
+    ngx_http_upstream_rr_peer_t **rpeer)
+{
+    ngx_int_t                      total;
+    ngx_uint_t                     i, p;
+    ngx_http_upstream_rr_peer_t  *peer, *best;
+
+    best = NULL;
+    p = 0;
+    total = 0;
+    for (peer = peers->peer, i = 0; peer; peer = peer->next, i++) {
+        peer->current_weight += peer->effective_weight;
+        total += peer->effective_weight;
+
+        if (best == NULL || peer->current_weight > best->current_weight) {
+            best = peer;
+            p = i;
+        }
+    }
+
+    *rpeer = best;
+    if (best == NULL) {
+        return NGX_ERROR;
+    }
+
+    best->current_weight -= total;
+
+    return p;
+}
+
+
+static ngx_http_upstream_rr_peer_t *
+ngx_http_upstream_get_vnswrr(ngx_http_upstream_vnswrr_peer_data_t  *vnsp)
+{
+    time_t                                  now;
+    uintptr_t                               m;
+    ngx_int_t                               rindex;
+    ngx_uint_t                              i, n, p, flag, begin_number;
+    ngx_http_upstream_rr_peer_t            *peer, *best;
+    ngx_http_upstream_rr_peers_t           *peers;
+    ngx_http_upstream_rr_vpeers_t          *vpeers;
+    ngx_http_upstream_rr_peer_data_t       *rrp;
+    ngx_http_upstream_vnswrr_srv_conf_t    *uvnscf;
+
+    now = ngx_time();
+
+    best = NULL;
+
+#if (NGX_SUPPRESS_WARN)
+    p = 0;
+#endif
+
+    rrp = &vnsp->rrp;
+    peers = rrp->peers;
+    uvnscf = vnsp->uvnscf;
+    vpeers = uvnscf->vpeers;
+
+    if (uvnscf->last_number == NGX_CONF_UNSET_UINT) {
+        uvnscf->init_number = ngx_random() % peers->number;
+
+        if (peers->weighted) {
+            peer = vpeers[uvnscf->init_number].vpeer;
+
+        } else {
+            for (peer = peers->peer, i = 0; i < uvnscf->init_number; i++) {
+                peer = peer->next;
+            }
+        }
+
+        uvnscf->last_number = uvnscf->init_number;
+        uvnscf->last_peer = peer;
+    }
+
+    if (peers->weighted) {
+        /* batch initialization vpeers at runtime. */
+        if (uvnscf->vnumber != peers->total_weight
+            && (uvnscf->last_number + 1 == uvnscf->vnumber))
+        {
+            n = peers->total_weight - uvnscf->vnumber;
+            if (n > peers->number) {
+                n = peers->number;
+            }
+
+            for (i = uvnscf->vnumber; i < n + uvnscf->vnumber; i++) {
+                rindex = ngx_http_upstream_get_rr_peer(peers, &peer);
+                if (rindex == NGX_ERROR) {
```

Comment:
如果`ngx_http_upstream_get_rr_peer ` 不会返回 `NGX_ERROR`，这里的判断是否也可以去掉？

Reply:
@jl2005  It's necessary, I think we should not rely on the implementation of the external API and discard the detection, because the implementation logic of the API may change at any time. 

Thx.